### PR TITLE
builtin/read: implement basic read builtin

### PR DESCRIPTION
**Pull request by delthas**

### CODE REVIEW

File: builtin/read.c [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	// TODO splitting into multiple variables with IFS
+	// TODO regular mode (no `-r`)
+	// assumes `-r`
+	if (argc > 2) {
+		fprintf(stderr, "Too many args for read\n");
+		return EXIT_FAILURE;
+	} else if (argc < 2) {
+		fprintf(stderr, "Not enough args for read\n");
+		return EXIT_FAILURE;
+	}
```
Comment (SirCmpwn) [inline builtin/read.c]: This should follow the same conventions for usage as the other builtins

File: builtin/read.c [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	size_t len = 0;
+	errno = 0;
+	ssize_t n = getline(&line, &len, stdin);
+	if (errno) {
+		free(line);
+		fprintf(stderr, "getline() failed: %s\n", strerror(errno));
+		return -1;
+	}
+	if (n < 0) {
+		n = strlen(line);
```
[the remark this answers is not in the document]
Comment (delthas) [reply builtin/read.c]: if the stream eofs, getline returns -1, so we have to get the length of the line ourselves

File: builtin/read.c [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+// usage: read [-r] var...
+static const char read_usage[] = "usage: read [-r] var...\n";
+
+int builtin_read(struct mrsh_state *state, int argc, char *argv[]) {
+	int raw = 0;
+	
+	int opt;
+	while ((opt = getopt(argc, argv, "r")) != -1) {
+		switch (opt) {
+			case 'r':
```
Comment (emersion) [inline builtin/read.c]: Style: can you indent `case` keywords with the same level as the `switch`?

```c
switch (thing) {
case 'x':
	do_x();
	break;
case 'y':
	do_y();
	break;
}
```

File: builtin/read.c [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+#include <unistd.h>
+#include <mrsh/buffer.h>
+#include <mrsh/shell.h>
+#include "builtin.h"
+
+// usage: read [-r] var...
+static const char read_usage[] = "usage: read [-r] var...\n";
+
+int builtin_read(struct mrsh_state *state, int argc, char *argv[]) {
+	int raw = 0;
```
Comment (emersion) [inline builtin/read.c]: This can be a `bool` (from `stdbool.h`)

File: builtin/read.c [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+		}
+	}
+	if (optind >= argc) {
+		fprintf(stderr, read_usage);
+		return EXIT_FAILURE;
+	}
+
+	// TODO handle multiple variables and split
+	if (argc > optind + 1) {
+		fprintf(stderr, "Too many args for read\n");
```
Comment (emersion) [inline builtin/read.c]: This can just be "read: multiple variables not yet implemented"

File: builtin/read.c [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+		return EXIT_FAILURE;
+	}
+	
+	struct mrsh_buffer buf = {0};
+	int escaped = 0;
```
Comment (emersion) [inline builtin/read.c]: This can be a `bool`

File: builtin/read.c [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+		}
+		if (c == '\n') {
+			if (escaped) {
+				escaped = 0;
+				continue;
+			}
+			break;
+		}
+		escaped = 0;
+		mrsh_buffer_append_char(&buf, (char) c);
```
Comment (emersion) [inline builtin/read.c]: Style: no space when casting: `(char)c`

**Comment by emersion** [comment]
Thanks!

**Pull request merged**